common/persistence/cassandra/nexus_service_store.go

```diff
@@ -61,30 +97,195 @@ func (s *NexusServiceStore) Close() {
+			strings.Join(columns, ","))
+	}
+
+	return nil
+}
+
 func (s *NexusServiceStore) GetNexusIncomingService(
 	ctx context.Context,
-	name string,
-) (*p.InternalGetNexusIncomingServiceResponse, error) {
-	return nil, serviceerror.NewUnimplemented("Cassandra GetNexusIncomingService is not implemented")
+	serviceID string,
+) (*p.InternalNexusIncomingService, error) {
+	query := s.session.Query(templateGetIncomingServiceQuery,
+		constServicesPartition,
+		rowTypeIncomingNexusService,
+		serviceID,
+	).WithContext(ctx)
+
+	var version int64
+	var dataBytes []byte
+	var dataEncoding string
+
+	if err := query.Scan(&dataBytes, &dataEncoding, &version); err != nil {
+		return nil, gocql.ConvertError("GetNexusIncomingService", err)
+	}
+
+	return &p.InternalNexusIncomingService{
+		ServiceID: serviceID,
+		Version:   version,
+		Data:      p.NewDataBlob(dataBytes, dataEncoding),
+	}, nil
 }
 
 func (s *NexusServiceStore) ListNexusIncomingServices(
 	ctx context.Context,
 	request *p.InternalListNexusIncomingServicesRequest,
 ) (*p.InternalListNexusIncomingServicesResponse, error) {
-	return nil, serviceerror.NewUnimplemented("Cassandra ListNexusIncomingServices is not implemented")
-}
+	response := &p.InternalListNexusIncomingServicesResponse{}
 
-func (s *NexusServiceStore) CreateOrUpdateNexusIncomingService(
-	ctx context.Context,
-	request *p.InternalCreateOrUpdateNexusIncomingServiceRequest,
-) error {
-	return serviceerror.NewUnimplemented("Cassandra CreateOrUpdateNexusIncomingService is not implemented")
+	query := s.session.Query(templateListIncomingServicesQuery, constServicesPartition, rowTypeIncomingNexusService).WithContext(ctx)
+	iter := query.PageSize(request.PageSize).PageState(request.NextPageToken).Iter()
+
+	row := make(map[string]interface{})
+	for iter.MapScan(row) {
+		serviceID, err := getTypedFieldFromRow[string]("service_id", row)
+		if err != nil {
+			return nil, err
+		}
+		version, err := getTypedFieldFromRow[int64]("version", row)
+		if err != nil {
+			return nil, err
+		}
+		data, err := getTypedFieldFromRow[[]byte]("data", row)
+		if err != nil {
+			return nil, err
+		}
+		dataEncoding, err := getTypedFieldFromRow[string]("data_encoding", row)
+		if err != nil {
+			return nil, err
+		}
+
+		response.Services = append(response.Services, p.InternalNexusIncomingService{
+			ServiceID: serviceID,
+			Version:   version,
+			Data:      p.NewDataBlob(data, dataEncoding),
+		})
+
+		row = make(map[string]interface{})
+	}
+
+	if len(iter.PageState()) > 0 {
+		response.NextPageToken = iter.PageState()
+	}
+
+	if err := iter.Close(); err != nil {
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf("ListNexusIncomingServices operation failed. Error: %v", err))
+	}
+
+	currentTableVersion, err := s.getTableVersion(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	response.TableVersion = currentTableVersion
+
+	if request.LastKnownTableVersion != UnknownTableVersion && request.LastKnownTableVersion != currentTableVersion {
```

Comment:
My problem with this approach is that the table version is inconsistent with the list result.
It works if the last known table version is not 0, otherwise, the list result may reflect an older version of the table.

For the first page, you actually want to list all rows including the partition state row. For the other pages, what you have here works.

Reply:
My concern is that I'm not sure we can guarantee that the table version row will be in the first page. I'm still pretty new to CQL though, so if we can be sure of this, then I can add the special handling for the first page.

Reply:
Also, I think as it is written now, we would only return an inconsistent list result if an incorrect table version was provided in the request and then between getting the results and getting the current table version, the table was updated so that now that originally incorrect table version is the current table version. For example: table version is 4, request comes in with version=5, get result page, table gets updated so version=5, then we check the current version and the condition passes. Otherwise if there was an update in between the two selects, this condition will fail and the caller will have to refetch the page with the new version.

Reply:
I think you can guarantee that and worst case add an order by clause.